### ISSUE TITLE
Leak in RCTNetwork module

### ISSUE THREAD

**Issue by yinghang** [facebook/react-native#19748]
- [x] Review the documentation: https://facebook.github.io/react-native
  - [x] Search for existing issues: https://github.com/facebook/react-native/issues
  - [x] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment
```
Environment:
  OS: macOS High Sierra 10.13.5
  Node: 6.11.1
  Yarn: 1.6.0
  npm: 3.10.10
  Watchman: 4.9.0
  Xcode: Not Found
  Android Studio: 2.3 AI-162.4069837

Packages: (wanted => installed)
  react: 16.3.1 => 16.3.1
  react-native: 0.55.4 => 0.55.4
```

## Description

A simple fetch call like the one below is enough to make various parts of `RCTNetwork` pile up references and memory allocations via `malloc` and etc. When I say pile up, what I mean is that every `GET` call is actually `malloc`-ing and retaining some data that never goes away. Simulating a low memory warning in Instruments also doesn't seem to get these references garbage collected away.

**Reproducibility: 100%**

```
  setInterval(() => {
    fetch('https://facebook.github.io/react-native/movies.json')
      .then((response) => response.json())
  }, 1000)
```

<details><summary>Screenshot 1</summary>

![screen shot 2018-06-13 at 4 16 21 pm](https://user-images.githubusercontent.com/4490352/41494445-14be696e-70e1-11e8-9311-ec7c4e634e0d.png)
</details>

<details><summary>Screenshot 2</summary>

<img width="1396" alt="screen shot 2018-06-15 at 9 25 24 pm" src="https://user-images.githubusercontent.com/4490352/41494567-dc137026-70e2-11e8-970c-1395dc7c11c9.png">
</details>

<details><summary>Screenshot 3</summary>

<img width="1396" alt="screen shot 2018-06-15 at 9 25 17 pm" src="https://user-images.githubusercontent.com/4490352/41494568-e18b3584-70e2-11e8-9c22-cc0dc87c69f8.png">
</details>

<details><summary>Screenshot 4</summary>

<img width="1396" alt="screen shot 2018-06-15 at 9 25 05 pm" src="https://user-images.githubusercontent.com/4490352/41494573-e88e1bb2-70e2-11e8-96da-7ca0a152a0b9.png">
</details>

Some of the more obvious locations that I observed to be leaking:
- [RCTNetworkTask.m](https://github.com/facebook/react-native/blob/master/Libraries/Network/RCTNetworkTask.m#L165) - There seems to be a retain cycle in the `completionBlock`
- [RCTNetworking.mm](https://github.com/facebook/react-native/blob/master/Libraries/Network/RCTNetworking.mm#L571)
- RCTHTTPRequestHandler.mm

**TLDR**: I'm not too familiar with the inner workings with `RCTNetwork` but the memory allocation seems to suggest that something funky is going on with the `completionBlock` which is leading to the retention of these references.

This is also a continuation of the conversation we had in #19169. @oNaiPs, you might be interested in this. 😄  

## Reproducible Demo
1. Clone this basic RN project @ https://github.com/yinghang/react-native-leak
2. Profile app with `Instruments` using `Allocations`
3. Watch `RCTNetwork` related references pile up

**Comment by ajwhite** [comment]
cc a couple folks that show up in the blame for the networking modules
@javache @nicklockwood

**Comment by Ashoat** [comment]
Pretty sure I'm seeing this on RN0.56 as well, though for some reason it's not showing up in the `Allocations` tool mentioned in the issue. `System Monitor` shows the memory usage grow to 1400 MiB and the device log shows the kernel sending the process a `SIGKILL` due to the hard memory limit being reached. Once, the app crashed on an `NSMallocException` in `RCTNetworkTask.m`.

So far I've been able to largely alleviate the issue by getting rid of the `Promise.race` that wraps my `fetch` call. Not sure to what degree I've been successful yet.

**Comment by Ashoat** [comment]
I've opened #20352 for my issue. It turns out downloading a specifically-formatted tiny JSON file can cause a much bigger memory leak.

Separately, I'm definitely able to confirm that OP's issue is still occurring in 0.56.0. The memory leak is much smaller, but still significant.

**Comment by corpsemxq** [comment]
Will this be fixed by https://github.com/facebook/react-native/pull/19333?
https://github.com/facebook/react-native/commit/122b3791ede095345f44666691aa9ce5aa7f725a

**Comment by yinghang** [comment]
Not too sure, will def give it a try when I have time.
